### PR TITLE
get private group info by group name

### PR DESCRIPTION
**Pull request by hussamEddinAlhendi**
@RocketChat/core 

add option in groups.info api to get group info by group name

### CODE REVIEW

File: packages/rocketchat-api/server/v1/groups.js [hunk 1]
```diff
@@ -227,15 +244,20 @@ RocketChat.API.v1.addRoute('groups.history', { authRequired: true }, {
...
+		}
 
 		//The find method returns either with the group or the failure
 		if (findResult.statusCode) {
 			return findResult;
 		}
 
 		return RocketChat.API.v1.success({
-			group: RocketChat.models.Rooms.findOneById(findResult.rid, { fields: RocketChat.API.v1.defaultFieldsToExclude })
+			group: RocketChat.models.Rooms.findOneByName(findResult.name, { fields: RocketChat.API.v1.defaultFieldsToExclude })
```
Comment (graywolf336) [inline packages/rocketchat-api/server/v1/groups.js]: This changed isn't required, I'd like this to be reset back to how it was.

File: packages/rocketchat-lib/server/models/Subscriptions.coffee [hunk 2]
```diff
@@ -19,19 +19,29 @@ class ModelSubscriptions extends RocketChat.models._Base
...
 		this.cache.ensureIndex(['rid', 'u._id'], 'unique')
+		this.cache.ensureIndex(['name', 'u._id'], 'unique')
 
 
 	# FIND ONE
 	findOneByRoomIdAndUserId: (roomId, userId) ->
 		if this.useCache
 			return this.cache.findByIndex('rid,u._id', [roomId, userId]).fetch()
-
+
```
Comment (graywolf336) [inline packages/rocketchat-lib/server/models/Subscriptions.coffee]: Remove this added space, thanks.

File: packages/rocketchat-api/server/v1/groups.js [hunk 3]
```diff
@@ -16,6 +16,23 @@ function findPrivateGroupById({ roomId, userId, checkedArchived = true }) {
 
 	return roomSub;
 }
+function findPrivateGroupByName({ roomName, userId, checkedArchived = true }) {
```
Comment (graywolf336) [inline packages/rocketchat-api/server/v1/groups.js]: Please combine these two functions into one. `findPrivateGroupByIdOrName({ roomId, roomName, userId, checkedArchived })`

File: packages/rocketchat-api/server/v1/groups.js [hunk 4]
```diff
@@ -227,15 +244,20 @@ RocketChat.API.v1.addRoute('groups.history', { authRequired: true }, {
 
 RocketChat.API.v1.addRoute('groups.info', { authRequired: true }, {
 	get: function() {
-		const findResult = findPrivateGroupById({ roomId: this.queryParams.roomId, userId: this.userId, checkedArchived: false });
+		let findResult;
+		if (this.queryParams.roomId) {
+			findResult = findPrivateGroupById({ roomId: this.queryParams.roomId, userId: this.userId, checkedArchived: false });
+		} else if (this.queryParams.roomName) {
```
Comment (graywolf336) [inline packages/rocketchat-api/server/v1/groups.js]: After the two functions are combined, combine them here as well to be one call.

File: packages/rocketchat-api/server/v1/groups.js [hunk 5]
```diff
@@ -1,10 +1,15 @@
 //Returns the private group subscription IF found otherwise it will reutrn the failure of why it didn't. Check the `statusCode` property
-function findPrivateGroupById({ roomId, userId, checkedArchived = true }) {
-	if (!roomId || !roomId.trim()) {
+function findPrivateGroupByIdOrName({ roomId, roomName, userId, checkedArchived = true }) {
+	if ((!roomId || !roomId.trim()) && (!roomName || !roomName.trim())) {
 		return RocketChat.API.v1.failure('Body param "roomId" is required');
```
Comment (graywolf336) [inline packages/rocketchat-api/server/v1/groups.js]: Thanks for making the changes, however this message now needs to be updated since either `roomId` or `roomName` are required.

Comment (hussamEddinAlhendi) [inline packages/rocketchat-api/server/v1/groups.js]: I pulled the latest code and this line has been changed to :
`throw new Meteor.Error('error-roomid-param-not-provided', 'The parameter "roomId" is required');`

File: packages/rocketchat-api/server/v1/groups.js [hunk 6]
```diff
@@ -430,7 +435,16 @@ RocketChat.API.v1.addRoute('groups.setType', { authRequired: true }, {
...
+<<<<<<< HEAD
+		const findResult = findPrivateGroupByIdOrName({ roomId: this.bodyParams.roomId, userId: this.userId });
+
+		//The find method returns either with the group or the failure
+		if (findResult.statusCode) {
+			return findResult;
+		}
+=======
 		const findResult = findPrivateGroupById({ roomId: this.bodyParams.roomId, userId: this.userId, checkedArchived: false });
+>>>>>>> 0d6d3f546aa18bc0778b367186c13fb45ba1d6c2
```
Comment (graywolf336) [inline packages/rocketchat-api/server/v1/groups.js]: Please fix this merge conflict and test before you commit, that way you know it works and builds.

File: packages/rocketchat-api/server/v1/channels.js [hunk 7]
```diff
@@ -1,7 +1,7 @@
 //Returns the channel IF found otherwise it will return the failure of why it didn't. Check the `statusCode` property
 function findChannelById({ roomId, checkedArchived = true }) {
 	if (!roomId || !roomId.trim()) {
-		throw new Meteor.Error('error-roomid-param-not-provided', 'The parameter "roomId" is required');
+		throw new Meteor.Error('error-roomid-param-not-provided', 'The parameter "roomId" or "roomName" is required');
```
Comment (graywolf336) [inline packages/rocketchat-api/server/v1/channels.js]: Channels still requires only the `roomId` and not `roomName`, please revert this change and move it to the groups. :+1:

Comment (hussamEddinAlhendi) [inline packages/rocketchat-api/server/v1/channels.js]: sorry I made mistake

Comment (graywolf336) [inline packages/rocketchat-api/server/v1/channels.js]: We all make mistakes, no worries! :)

File: packages/rocketchat-api/server/v1/groups.js [hunk 8]
```diff
@@ -435,16 +435,7 @@ RocketChat.API.v1.addRoute('groups.setType', { authRequired: true }, {
...
 	post: function() {
-<<<<<<< HEAD
-		const findResult = findPrivateGroupByIdOrName({ roomId: this.bodyParams.roomId, userId: this.userId });
-
-		//The find method returns either with the group or the failure
-		if (findResult.statusCode) {
-			return findResult;
-		}
-=======
 		const findResult = findPrivateGroupById({ roomId: this.bodyParams.roomId, userId: this.userId, checkedArchived: false });
```
Comment (graywolf336) [inline packages/rocketchat-api/server/v1/groups.js]: You missed the change here I believe, shouldn't it be using `findPrivateGroupByIdOrName` now?

Comment (hussamEddinAlhendi) [inline packages/rocketchat-api/server/v1/groups.js]: because I merged the conflicts and forgot to do changes

**Comment by graywolf336** [comment]
Can you send a screenshot of the methods working?

**Comment by hussamEddinAlhendi** [comment]
screen shot 2017-01-16 at 7 15 26 pm

File: packages/rocketchat-lib/server/models/Subscriptions.coffee [hunk 9]
```diff
@@ -19,19 +19,28 @@ class ModelSubscriptions extends RocketChat.models._Base
 		this.cache.ensureIndex('rid', 'array')
 		this.cache.ensureIndex('u._id', 'array')
 		this.cache.ensureIndex(['rid', 'u._id'], 'unique')
+		this.cache.ensureIndex(['name', 'u._id'], 'unique')
```
Comment (graywolf336) [inline packages/rocketchat-lib/server/models/Subscriptions.coffee]: @rodrigok Thoughts on this line?

Comment (rodrigok) [inline packages/rocketchat-lib/server/models/Subscriptions.coffee]: @graywolf336 That is ok. The room name is unique, so room name + user id will be unique for subscriptions
